src/python/WMCore/JobSplitting/EventAwareLumiByWork.py

```diff
@@ -0,0 +1,276 @@
+        elif runs and lumis:
+            lumiMask = LumiList(wmagentFormat=(runs, lumis))
+        elif runWhitelist:
+            lumiMask = LumiList(runs=runWhitelist)
+
+        # EWV Not looked at yet
+        self.perfParameters = kwargs.get('performance', {})
+        eventsPerLumiInDataset = 0
+
+        if deterministicPileup and self.package == 'WMCore.WMBS':
+            getJobNumber = self.daoFactory(classname="Jobs.GetNumberOfJobsPerWorkflow")
+            jobNumber = getJobNumber.execute(workflow=self.subscription.getWorkflow().id)
+            self.nJobs = jobNumber
+        # EWV End not looked at yet
+
+        # If we have runLumi info, we need to load it from couch
+        if collectionName:
+            lumiMask = self.lumiMaskFromACDC(couchURL=kwargs.get('couchURL'), couchDB=kwargs.get('couchDB'),
+                                             filesetName=kwargs.get('filesetName'), collectionName=collectionName,
+                                             owner=kwargs.get('owner'), group=kwargs.get('group'))
+
+        lDict = self.sortByLocation()
+
+        # First we need to load the data
+        if self.package == 'WMCore.WMBS':
+            loadRunLumi = self.daoFactory(classname="Files.GetBulkRunLumi")
+
+        # Just gather up which runs and lumis are in the dataset for each location
+        # defaultdict of defaultdict will create keys on reference that don't exist
+        eventsByLumi = defaultdict(lambda: defaultdict(float))
+        filesByLumi = defaultdict(lambda: defaultdict(list))
+        lumisByLocation = {}
+        lumisToProcess = LumiList()
+        # TODO: Do I really need to use strings for dictionary keys?
+        for locationSet in lDict.keys():
+            # Ignore WMCore.WMBS case for now
+            runsAndLumis = []
+            for fileObj in lDict[locationSet]:
+                eventsInFile = fileObj['events']
+                lumisInFile = self.countLumis(fileObj['runs'])
+                if getParents:
+                    parentLFNs = self.findParent(lfn=fileObj['lfn'])
+                    for lfn in parentLFNs:
+                        parent = File(lfn=lfn)
+                        fileObj['parents'].add(parent)
+
+                for runLumi in fileObj['runs']:
+                    run = str(runLumi.run)
+                    runsAndLumis.append({run: runLumi.lumis})  # Accumulate a list of runs and lumis
+                    for lumiNum in runLumi.lumis:
+                        lumi = str(lumiNum)
+                        filesByLumi[run][lumi].append(fileObj)
+                        eventsByLumi[run][lumi] = round(eventsInFile / lumisInFile)
+
+            # Use fast construction to build a LumiList for the FileSet and apply the mask determined earlier
+            lumisByLocation[locationSet] = LumiList(runsAndLumis=runsAndLumis)
+            if lumiMask:
+                lumisByLocation[locationSet] &= lumiMask
+            lumisToProcess += lumisByLocation[locationSet]
```

Comment:
Wouldn't it be better to apply the mask earlier, before you build up the full filesByLumi and eventsByLumi dictionaries. If you do sparse processing of large filesets this can make quite a difference...